src/python/grpcio/grpc/_channel.py

```diff
@@ -388,65 +478,86 @@ def _next(self):
                     elif self._state.code is not None:
                         raise self
 
-    def __next__(self):
-        return self._next()
+    def _next(self):
+        with self._state.condition:
+            if self._state.code is None:
+                operating = self._call.operate(
+                    (cygrpc.ReceiveMessageOperation(_EMPTY_FLAGS),), None)
+                if operating:
+                    self._state.due.add(cygrpc.OperationType.receive_message)
+            elif self._state.code is grpc.StatusCode.OK:
+                raise StopIteration()
+            else:
+                raise self
+        return self._next_response()
 
-    def next(self):
-        return self._next()
+    def debug_error_string(self):
+        with self._state.condition:
+            if self._state.debug_error_string is None:
+                raise RuntimeError(
+                    "Cannot get debug error string until RPC is completed.")
+            return _common.decode(self._state.debug_error_string)
 
-    def __iter__(self):
-        return self
 
-    def debug_error_string(self):
+class _MultiThreadedRendezvous(_Rendezvous, grpc.Future):  # pylint: disable=too-many-ancestors
+    """An RPC iterator that depends on a channel spin thread.
+
+    This iterator relies upon a per-channel thread running in the background,
+    dequeueing events from the completion queue, and notifying threads waiting
+    on the threading.Condition object in the _RPCState object.
+
+    This extra thread allows _MultiThreadedRendezvous to fulfill the grpc.Future interface
+    and to mediate a bidirection streaming RPC.
+    """
+
+    def initial_metadata(self):
+        """See grpc.Call.initial_metadata"""
         with self._state.condition:
 
             def _done():
-                return self._state.debug_error_string is not None
+                return self._state.initial_metadata is not None
 
             _common.wait(self._state.condition.wait, _done)
-            return _common.decode(self._state.debug_error_string)
+            return self._state.initial_metadata
 
-    def _repr(self):
+    def trailing_metadata(self):
+        """See grpc.Call.trailing_metadata"""
         with self._state.condition:
-            if self._state.code is None:
-                return '<{} object of in-flight RPC>'.format(
-                    self.__class__.__name__)
-            elif self._state.code is grpc.StatusCode.OK:
-                return _OK_RENDEZVOUS_REPR_FORMAT.format(
-                    self._state.code, self._state.details)
-            else:
-                return _NON_OK_RENDEZVOUS_REPR_FORMAT.format(
-                    self._state.code, self._state.details,
-                    self._state.debug_error_string)
 
-    def __repr__(self):
-        return self._repr()
+            def _done():
+                return self._state.trailing_metadata is not None
```

Comment:
Do you think these `_done` functions can be shorten into `lambda`s?

Reply:
Personally, I think Python lambdas make code *less* readable. Their syntax is too similar to Python typing.